src/net/larse/lcms/algorithms/VCT.java

```diff
@@ -0,0 +1,1567 @@
+        // Continue over bad pixels or no detected disturbances
+        if (this.qFlag[i] == QA_BAD || isMinorFirst(i) == false) {
+          i++;
+          continue;
+        }
+
+        // Increment j as long as minor disturbance is detected
+        // (is_minor_rest == 1).  If the year range is at least two years,
+        // characterize it as a disturbance
+        int j = i + 1;
+        while (j <= endYear && this.qFlag[j] == QA_GOOD
+                && isMinorRest(j) == true) {
+          j++;
+        }
+        if (j - i > 1) {
+          setDisturbance(i, j);
+        }
+        i = j;
+      }
+    }
+
+    /**
+     * Calculate the change magnitude (in many indexes) for a given disturbance
+     * segment
+     *
+     * @param changeYear - year of initial disturbance onset
+     * @param peakYear - year of peak disturbance
+     */
+    private void calculateChangeMagn(int changeYear, int peakYear) {
+      // Handle exception
+      // if (changeYear < 0 || changeYear >= this.numYears
+      //         || peakYear < 0 || peakYear >= this.numYears) {
+      //   String msg = "Year out of range in calculateChangeMagn";
+      //   throw new Exception(msg);
+      // }
+
+      // Capture the disturbance magnitudes in UD, NDVI and DNBR spaces.
+      // Magnitudes are relative from the peak UD year to the mean of
+      // the pixel across all years
+      this.distMagn[changeYear]
+              = this.ud[COMP][peakYear] - this.meanForUdBx[COMP];
+      this.distMagnB4[changeYear]
+              = this.ud[B4][peakYear] - this.meanForUdBx[B4];
+      this.distMagnVi[changeYear]
+              = this.ud[NDVI][peakYear] - this.meanForUdBx[NDVI];
+      this.distMagnBr[changeYear]
+              = this.ud[DNBR][peakYear] - this.meanForUdBx[DNBR];
+    }
+
+    /**
+     * Calculate a linear change rate in ud B5 over a given time span using a
+     * simple linear fit
+     *
+     * @param startYear - start year
+     * @param endYear - end year
+     * @return - Regression coefficients from least squares fit
+     */
+    private double[] calculateLinearChangeRate(int startYear, int endYear) {
+      // Too few observations for meaningful regression, or not much
+      // variability. Return default values
+      if (endYear - startYear < 4 || this.fiRange < 0.1) {
+        return FALSE_FIT;
+      }
+
+      // Index errors on start_year, end_year
+      // if (endYear > this.numYears || startYear < 0) {
+      //   // Handle exception
+      //   String msg = "Error: index error in calculateLinearChangeRate";
+      //   throw new Exception(msg);
+      // }
+
+      if (endYear == this.numYears) {
+        endYear--;
+      }
+
+      // Linear regression of B5 (Y) on years since disturbance (X)
+      // TODO: Reimplement using LinearLeastSquares - for now using 
+      // Cheng's code for consistency
+      //
+      // TODO: Generalize this function for any index
+      int i, j;
+      double[] x = new double[this.numYears];
```

Comment:
This function is called numYears times. I think that means x and y should be member variables so they're not allocated and deallocated (30 * nPixels) times.